### ISSUE TITLE
Add TER metric 

### ISSUE THREAD

**Issue by alvations** [mjpost/sacrebleu#90]
How do we go about contributing new metrics to SacreBLEU? 

There's the implementation of TER on https://gist.github.com/alvations/989dacca76c820c48ef18c4637515b39

**Comment by ozancaglayan** [comment]
Hi,

I have PR request  #88 which isolates the metrics into their own classes. If that passes, we can add TER, METEOR and potentially CIDEr and ROUGE from COCO metrics if @mjpost agrees to extend this project something like a sacreval rather than sacreBLEU. This will potentially benefit also from upcoming statistical significance testing for all these metrics. If not, I (we) may also create another project which aggregates all these stuff.

**Comment by mjpost** [comment]
I think this is a nice idea, and love the name `sacreval`. I wonder if it's best to create a new repo / tool that uses `sacrebleu` as a package for the BLEU implementation. I'd be happy to be involved but in a limited capacity.

I would also recommend that you create reference implementations, that implement an established variant. This helps adoption of your tool and is consistent with sacrebleu's goal of ensuring numbers across papers are comparable.

**Comment by martinpopel** [comment]
> I wonder if it's best to create a new repo / tool that uses sacrebleu as a package for the BLEU implementation

I am afraid that it would still mean a lot of code duplication. Currently a lot of logic is implemented directly in `main()`. For example the handling of `--origlang` and `-t` accepting a comma-separated list of test sets is not supported by the API, it is only available from the command line.
Also, the signature is [difficult to get via the API](https://github.com/mjpost/sacrebleu/issues/75) and the `sacreval` signatures would probably need to contain both the version of `sacrebleu` and of `sacreval` itself.

What about implementing the additional metrics in this repo and one of the following:
1. Don't introduce the `sacreval` name at all. I like the `sacreval` name as well*, but it `sacrebleu` has already gained some popularity. It seems everyone understands that it can compute `chrF`, so they should not be surprised with `sacrebleu -m TER`.
2. Keep `mjpost/sacrebleu` as the repo name and `sacrebleu` as thy PyPI package, but add `sacreval` as an alias for `sacrebleu` binary.
3. Rename the repo and the PyPI package to `sacreval` and make `sacrebleu` a legacy alias for the `sacreval` binary. Promote the transition to the new name. We could keep the `sacrebleu` PyPI package, but implement it as an wrapper around `sacreval`.

My preferences are 1, 2, 3, in this order,

`sacreval` as a new repo may be also an option, but only after further sacreBLEU refactoring to prevent code duplication. Such refactoring may be a good idea anyway*.

*) I implemented a tool called `t2t-sacreval`, which works as [t2t-bleu](https://github.com/tensorflow/tensor2tensor/blob/master/tensor2tensor/bin/t2t_bleu.py) (it generates files for TensorBoard with BLEU learning curves) but using sacreBLEU instead of the T2T's BLEU implementation. I have not published this tool because it supports only a subset of `sacrebleu`'s options and the implementation is ugly, just for my needs.

**Comment by ozancaglayan** [comment]
sacreval was just a suggestion, not so important at this point :) Here are
some random/unorganized thoughts:

We could however factorize and modularize the codebase to automatically
provide per-metric scripts i.e. sacre-bleu, sacre-chrf, sacre-meteor etc.

A holistic sacrebleu(or sacreval) can still exist and aggregate
functionalities (subparser from argparse can potentially do this) from the
available/installed modules.

If one particular submodule is not available because of missing
dependencies etc. we can be graceful and disable metrics at runtime.

I don't know if the aforementioned TER implementation relies on the
reference Java implementation or not. If so, interfacing Java and offering
TER JAR (and potentially Meteor JARs) should also be discussed.

In short, if we can factorize everything out and let the metrics provide a
unified API, new metrics will easily be pluggable without bottering the
core functionalities. sacrebleu can go on as it is, to provide chrf and
bleu (mostly for backward compatibility) along with new script(s) extending
the metric inventory.
